core/frontend/src/render/webgl/Texture.ts

```diff
@@ -480,6 +511,88 @@ export class Texture2DHandle extends TextureHandle {
   }
 }
 
+/** @internal */
+export interface ExternalTextureRequest {
+  handle: Texture2DHandle;
+  name: string;
+  imodel: IModelConnection;
+  type: RenderTexture.Type;
+  format: ImageSourceFormat;
+  onLoaded?: (req: ExternalTextureRequest) => void;
+}
+
+/** @internal */
+export class ExternalTextureLoader { /* currently exported for tests only */
+  public static readonly instance = new ExternalTextureLoader(10);
+  private readonly _maxActiveRequests: number;
+  private _activeRequests: Array<ExternalTextureRequest> = [];
+  private _pendingRequests: Array<ExternalTextureRequest> = [];
+
+  public get numActiveRequests() { return this._activeRequests.length; }
+  public get numPendingRequests() { return this._pendingRequests.length; }
+  public get maxActiveRequests() { return this._maxActiveRequests; }
+
+  private constructor(maxActiveRequests: number) {
+    this._maxActiveRequests = maxActiveRequests;
+  }
+
+  private _nextRequest(prevReq: ExternalTextureRequest) {
+    this._activeRequests.splice(this._activeRequests.indexOf(prevReq), 1);
+    if (this._activeRequests.length < this._maxActiveRequests && this._pendingRequests.length > 0) {
+      const req = this._pendingRequests.pop()!;

```

Comment:
Is this intended to be a first-in last-out queue?

Reply:
Not necessarily - array has a nice, easy .pop() method, but yeah, it's not FIFO then.  Think we should make it that way?

Reply:
Seems like what you'd want.

Reply:
This uses .shift() now to emulate a queue.